### PR TITLE
Feature/sap system facts

### PR DESCRIPTION
**Pull request by rainerleber**
##### SUMMARY
New facts module which gathers facts about the running system

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
sap_system_facts

##### ADDITIONAL INFORMATION
-

### CODE REVIEW

File: plugins/modules/system/sap_system_facts.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: sap_system_facts
+
+short_description: Gathers SAP facts in a host
+
+version_added: "0.2.0"
```
Comment (rkpobe) [inline plugins/modules/system/sap_system_facts.py:14]: should this be version "0.1.0"?

File: plugins/modules/system/sap_system_facts.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+RETURN = r'''
+# These are examples of possible return values, and in general should use other names for return values.
+ansible_facts:
+  description: Facts to add to ansible_facts.
+  returned: always
+  type: list
+  contains:
+    sap:
+      description: Facts about the running SAP system.
+      type: str
```
Comment (rkpobe) [inline plugins/modules/system/sap_system_facts.py]: type list contains dict with strings?

File: plugins/modules/system/sap_system_facts.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            command = [module.get_bin_path('/usr/sap/hostctrl/exe/sapcontrol', required=True)]
+            # check if returned instance_nr is a number because sapcontrol returns all if a random string is provided
+            if instance_nr.isdigit():
+                command.extend(['-nr', instance_nr, '-function', 'GetInstanceProperties'])
+                check_instance = module.run_command(command, check_rc=False)
+                if check_instance[0] != 1:
+                    for line in check_instance[1].splitlines():
+                        if re.search('INSTANCE_NAME', line):
+                            # convert to list and extract last
+                            type_raw = (line.strip('][').split(', '))[-1]
```
Comment (rkpobe) [inline plugins/modules/system/sap_system_facts.py:132]: line is e.g. from GetInstanceProperties (on a Hana system): "INSTANCE_NAME, Attribute, HDB01":
 How is this processed?
```
(line.strip('][').split(', '))[-1]) --> type_raw='HDB01'
--> type = '01' --> 
get_instance_type('01') --> 'TYPE': 'XXX' --> Unknown instance type.
``` 
Correct?

File: plugins/modules/system/sap_system_facts.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    type = ""
+    for sid in sids:
+        for instance in os.listdir('/usr/sap/' + sid):
+            instance_nr = instance[-2:]
+            command = [module.get_bin_path('/usr/sap/hostctrl/exe/sapcontrol', required=True)]
+            # check if returned instance_nr is a number because sapcontrol returns all if a random string is provided
+            if instance_nr.isdigit():
+                command.extend(['-nr', instance_nr, '-function', 'GetInstanceProperties'])
+                check_instance = module.run_command(command, check_rc=False)
+                if check_instance[0] != 1:
```
Comment (rkpobe) [inline plugins/modules/system/sap_system_facts.py:128]: would check_instance[0]==0 be more suitable? (Which return-code-cases besides 0 have been tested 2,3,4 ... n?)

File: plugins/modules/system/sap_system_facts.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    )
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        supports_check_mode=True
+    )
+
+    hana_sid = get_all_hana_sid()
+    if hana_sid:
+        system_result = system_result + get_hana_nr(hana_sid, module)
```
Comment (rkpobe) [inline plugins/modules/system/sap_system_facts.py:180]: technically hana_sid is not a boolean (same for nw_sid and system_result). But still works.

File: tests/unit/plugins/modules/system/test_sap_system_facts.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                                                                             {"InstanceType": "NW", "NR": "01", "SID": "ABC", "TYPE": "PAS"}]})
+
+    def test_sap_system_facts_command_hana(self):
+        """Check that result for HANA is correct."""
+        with patch.object(self.module, 'get_all_hana_sid') as mock_all_hana_sid:
+            mock_all_hana_sid.return_value = ['HDB']
+            with patch.object(self.module.os, 'listdir') as mock_listdir:
+                mock_listdir.return_value = ['HDB01']
+                with patch.object(basic.AnsibleModule, 'run_command') as run_command:
+                    run_command.return_value = [0, '', '']
```
Comment (rkpobe) [inline tests/unit/plugins/modules/system/test_sap_system_facts.py:67]: Why does this not result in [0,'INSTANCE_NAME, Attribute, HDB01',''"]?

File: plugins/modules/system/sap_system_facts.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        # It's an SCS
+        type = "SCS"
+    elif (raw_type[0] == "E"):
+        # It's an ERS
+        type = "ERS"
+    else:
+        # Unknown instance type
+        type = "XXX"
+    return type
+
```
Comment (rkpobe) [inline plugins/modules/system/sap_system_facts.py:162]: There is no return type "HDB"

**Review by rkpobe — COMMENTED** [review]
There are a few questions left regarding the parsing of the output from 'GetInstanceProperties'

File: plugins/modules/system/sap_system_facts.py (hunk 4 above)
Comment (rainerleber) [reply plugins/modules/system/sap_system_facts.py:128]: sapcontrol exit codes are betwenn 0 - 4 and n - 255

For example:
    0  Last webmethod call successful
    1  Last webmethod call failed, invalid parameter
    2  StartWait, StopWait, WaitforStarted, WaitforStopped, RestartServiceWait
       timed out
       CheckSystemCertificates detected warnings
    3  GetProcessList succeeded, all processes running correctly
       CheckSystemCertificates detected errors
    4  GetProcessList succeeded, all processes stopped

So this assumes if  exit is not 1 the instance number is available wether the system is down or not.

File: plugins/modules/system/sap_system_facts.py (hunk 1 above)
Comment (rainerleber) [reply plugins/modules/system/sap_system_facts.py:14]: next version would be 0.2.0 unitl then the module is not available on galaxy

**Review by rkpobe — APPROVED** [review]

**Pull request merged**